Speed up RecordFunction with sampled callbacks

ilia-cher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#20307 Allow lazy initialization of inputs in RecordFunction**

Summary:
In production environment we typically use sampling and don't
require expensive input copying on every function call, this PR
enables lazy inputs copying, when actual RecordFunction callback
calls inputs()

Test Plan:
BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake
./build/bin/test_jit --gtest_filter=JitTest.RecordFunction

Differential Revision: [D15276308](https://our.internmc.facebook.com/intern/diff/D15276308)

Commit:
Allow lazy initialization of inputs in RecordFunction

Summary:
In production environment we typically use sampling and don't
require expensive input copying on every function call, this PR
enables lazy inputs copying, when actual RecordFunction callback
calls inputs()

Test Plan:
iBLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop
--cmake
./build/bin/test_jit --gtest_filter=JitTest.RecordFunction

Commit:
Update on "Allow lazy initialization of inputs in RecordFunction"

Allow lazy initialization of inputs in RecordFunction

Summary:
In production environment we typically use sampling and don't
require expensive input copying on every function call, this PR
enables lazy inputs copying, when actual RecordFunction callback
calls inputs()

Test Plan:
iBLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop
--cmake
./build/bin/test_jit --gtest_filter=JitTest.RecordFunction

gh-metadata: pytorch pytorch 20307 gh/ilia-cher/17/head

Commit:
Update on "Allow lazy initialization of inputs in RecordFunction"

Allow lazy initialization of inputs in RecordFunction

Summary:
In production environment we typically use sampling and don't
require expensive input copying on every function call, this PR
enables lazy inputs copying, when actual RecordFunction callback
calls inputs()

Test Plan:
iBLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop
--cmake
./build/bin/test_jit --gtest_filter=JitTest.RecordFunction

gh-metadata: pytorch pytorch 20307 gh/ilia-cher/17/head

zdevito reviewed (commented):
Is there a downside to making lazy_inputs_copy always true? It's already under a guard so it doesn't effect the does not needs inputs case. I'd prefer not to add options unless there is a strong justification for keeping both.

ilia-cher commented:
@zdevito I think it is not strictly better than what we have now, e.g. we have use cases when we want inputs on every call, for example shape profiling

zdevito commented:
But the cost of making and evaluating the lambda is going to be almost nothing compared to the cost of generating the list of inputs, right? I just don't  like the idea of adding an option here to save almost nothing.

ilia-cher commented:
You're probably right about the overall trade-off between complexity and potential perf win here. I guess the only way to resolve this is to benchmark - will update with results. If the difference is small I will make it just always lazy.

zdevito commented:
I am willing to approve a diff that does lazy-by-default without any benchmarking since the pathway where we use sizes is already only used in profiling and only used rarely as of now. It's not going to break anything to add it. It is up to you whether you want to do additional benchmarking anyway.

Commit:
Update on "Speed up RecordFunction with sampled callbacks"

Speed up RecordFunction with sampled callbacks

Summary:
Reduce overhead of capturing inputs when using RecordFunction with
sampled callbacks

Test Plan:
iBLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop
--cmake
./build/bin/test_jit --gtest_filter=JitTest.RecordFunction

gh-metadata: pytorch pytorch 20307 gh/ilia-cher/17/head

ilia-cher commented:
Did the benchmarking for all cases.
The benchmarking patch is here: https://gist.github.com/ilia-cher/7fa02f3e05079d6d85c87de5e4c36662
Results: https://gist.github.com/ilia-cher/d4f461e42c7705833904a62ab00f1630

The conclusions:
- Lambda is not better than just always creating a vector. It's not surprising since we still need to create and save capture.
- Expectedly, lambda is also slower in case we do use inputs() (but that doesn't matter much since we were still going to call inputs() rarely, with sampling)

I've also included in the benchmark test cases: no callback at all, callback but without using any inputs.

So, we could work further on generalized version that (a) always captures inputs since we assume that callbacks use them on every call and (b) as efficient as no inputs used in case callbacks do not use inputs.

But, ultimately, we just don't want to waste resources in case callbacks don't use inputs and the only reasonable use case when they need inputs but don't use them is probabilistic sampling. One solution is to just move the decision point up into RECORD_FUNCTION - the current version of this PR does that.
Note that in this case, as expected, there's pretty much zero overhead when using callbacks with heavy sampling.

Commit:
Update on "Speed up RecordFunction with sampled callbacks"

Speed up RecordFunction with sampled callbacks

Summary:
Reduce overhead of capturing inputs when using RecordFunction with
sampled callbacks

Test Plan:
iBLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop
--cmake
./build/bin/test_jit --gtest_filter=JitTest.RecordFunction

gh-metadata: pytorch pytorch 20307 gh/ilia-cher/17/head

zdevito commented on torch/csrc/autograd/record_function.cpp:
```diff
@@ -40,6 +43,22 @@ bool needsInputs() {
   return callback_needs_inputs > 0;
 }
 
+bool isSampledCallbacks() {
```
I don't see the need for 4 functions here. Why not:

```
void setSamplingProbability(double prob); {
   if (prob == 1.0) {
     is_sampled_callbacks = false;   
  }
  AT_CHECK(prob > 0 && < 1.0);
  is_sabled_callbacks = true;
  sampling_prob = prob;
}
```

ilia-cher replied:
I guess if we just always use and pass constant 1.0 then == should work for doubles, although I wouldn't want to explicitly rely on this.. I can simplify interface, but I'd prefer to compare abs(diff) < epsilon

zdevito commented on torch/csrc/autograd/record_function.h:
```diff
@@ -92,14 +94,22 @@ struct TORCH_API RecordFunction {
-      guard.before(fn, inputs, ##__VA_ARGS__); \
-    } else { \
-      guard.before(fn, ##__VA_ARGS__); \
+    if (!torch::autograd::profiler::isSampledCallbacks() || \
```
Can you put:

```
!torch::autograd::profiler::isSampledCallbacks() || \
      guard.before(fn, inputs, ##__VA_ARGS__); \	        (((double) std::rand() / RAND_MAX) < torch::autograd::profiler::getSamplingProbability())
```

Into a (inlinable) function, so that it is easier to read what the macro is doing?

ilia-cher replied:
it's a bit messy because we have to have macro definition in header but can't have global variables in header, so the usage of TORCH_API seems inevitable, though we already call TORCH_API functions like hasCallbacks or needsInputs, so it shouldn't be a problem

ilia-cher replied:
or I could keep getSamplingProbability/isSampledCallbacks but I'd prefer to just have one `TORCH_API checkCallbacksEnabled` function

ilia-cher replied:
I think I can't make it inline because the its body needs to be in cpp file, or I need to keep other functions

zdevito reviewed (approved):
Sampling approach seems like a good way to go here. I just have a few comments to simplify the interface, and then it is ready to go.

Commit:
Update on "Speed up RecordFunction with sampled callbacks"

Speed up RecordFunction with sampled callbacks

Summary:
Reduce overhead of capturing inputs when using RecordFunction with
sampled callbacks

Test Plan:
iBLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop
--cmake
./build/bin/test_jit --gtest_filter=JitTest.RecordFunction

gh-metadata: pytorch pytorch 20307 gh/ilia-cher/17/head

Commit:
Update on "Speed up RecordFunction with sampled callbacks"

Speed up RecordFunction with sampled callbacks

Summary:
Reduce overhead of capturing inputs when using RecordFunction with
sampled callbacks

Test Plan:
iBLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop
--cmake
./build/bin/test_jit --gtest_filter=JitTest.RecordFunction

gh-metadata: pytorch pytorch 20307 gh/ilia-cher/17/head

Commit:
Update on "Speed up RecordFunction with sampled callbacks"

Speed up RecordFunction with sampled callbacks

Summary:
Reduce overhead of capturing inputs when using RecordFunction with
sampled callbacks

Test Plan:
iBLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop
--cmake
./build/bin/test_jit --gtest_filter=JitTest.RecordFunction

gh-metadata: pytorch pytorch 20307 gh/ilia-cher/17/head

Commit:
Update on "Speed up RecordFunction with sampled callbacks"

Speed up RecordFunction with sampled callbacks

Summary:
Reduce overhead of capturing inputs when using RecordFunction with
sampled callbacks

Test Plan:
iBLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop
--cmake
./build/bin/test_jit --gtest_filter=JitTest.RecordFunction

gh-metadata: pytorch pytorch 20307 gh/ilia-cher/17/head

Commit:
Update on "Speed up RecordFunction with sampled callbacks"

Speed up RecordFunction with sampled callbacks

Summary:
Reduce overhead of capturing inputs when using RecordFunction with
sampled callbacks

Test Plan:
iBLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop
--cmake
./build/bin/test_jit --gtest_filter=JitTest.RecordFunction

gh-metadata: pytorch pytorch 20307 gh/ilia-cher/17/head
